File: scripts/lib/svg.mjs
/**
 * Drawing primitives on a grid of terminal cells.
 *
 * Each grid row is `lh` tall. Regular text uses the font at its normal size,
 * with the baseline centered in the cell. Box-drawing and block characters
 * (`╭─│█▄`) use "tall" mode: the font is scaled up until the glyph fills the whole
 * cell and squeezed horizontally, so lines and blocks connect without gaps
 * between rows, like in a real terminal.
 */

export const n2 = (v) => Number(Number(v).toFixed(2));

export const esc = (s) => String(s)
  .replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;')
  .replace(/"/g, '&quot;').replace(/'/g, '&apos;');

export const fill = (template, vars) => String(template).replace(/\{(\w+)\}/g, (m, k) => (k in vars ? vars[k] : m));

export const len = (s) => [...String(s)].length;

export const pad = (s, width, align = 'left') => {
  const str = [...String(s)].slice(0, width).join('');
  const space = ' '.repeat(Math.max(0, width - len(str)));
  return align === 'right' ? space + str : str + space;
};

export function wrap(text, width) {
  const lines = [];
  for (const paragraph of String(text).split('\n')) {
    let line = '';
    for (const word of paragraph.split(/\s+/).filter(Boolean)) {
      if (!line) line = word;
      else if (len(line) + 1 + len(word) <= width) line += ` ${word}`;
      else { lines.push(line); line = word; }
      while (len(line) > width) { lines.push([...line].slice(0, width).join('')); line = [...line].slice(width).join(''); }
    }
    lines.push(line);
  }
  return lines;
}

const EIGHTHS = ['', '▏', '▎', '▍', '▌', '▋', '▊', '▉'];

export function createCanvas({ theme, layout, animate, k, glow }) {
  const fs = layout.fontSize;
  const cw = fs * 0.6;
  const lh = n2(fs * layout.lineHeight);
  const padX = 26;
  const titleH = 36;
  const originY = titleH + 16;
  const cols = Math.floor((layout.width - padX * 2) / cw);
  const tallSize = n2(lh / 1.32 + 0.15);
  const defs = [];
  let uid = 0;

  const c = {
    T: theme, fs, cw, lh, padX, titleH, originY, cols, animate, k, defs,
    width: layout.width,

    id: (prefix = 'i') => `${prefix}${uid++}`,
    X: (col) => n2(padX + col * cw),
    top: (row) => n2(originY + row * lh),
    Y: (row) => n2(originY + row * lh + (lh + 0.72 * fs) / 2),
    tallY: (row) => n2(originY + row * lh + 1.02 * tallSize - 0.08),
    at: (s) => `${n2(s)}s`,

    /** Soft phosphor glow around the content (when enabled). */
    glow: (inner) => (glow && inner ? `<g filter="url(#glow)">${inner}</g>` : inner),

    /** Text at pixel coordinates. */
    glyphs(x, y, str, color, { size = fs, weight, anchor, squeeze, opacity, seal } = {}) {
      const n = len(str);
      if (!n || !String(str).trim()) return '';
      const attrs = [`x="${n2(x)}"`, `y="${n2(y)}"`, `fill="${color}"`];
      const advance = squeeze ? cw : size * 0.6;
      if (n > 1 || squeeze) attrs.push(`textLength="${n2(n * advance)}"`, `lengthAdjust="${squeeze ? 'spacingAndGlyphs' : 'spacing'}"`);
      if (size !== fs) attrs.push(`style="font-size:${size}px"`);
      if (weight) attrs.push(`font-weight="${weight}"`);
      if (anchor) attrs.push(`text-anchor="${anchor}"`);
      if (opacity !== undefined) attrs.push(`fill-opacity="${opacity}"`);
      // "seal": a thin same-color stroke closes antialiasing gaps between blocks
      if (seal) attrs.push(`stroke="${color}"`, 'stroke-width="0.6"');
      return `<text ${attrs.join(' ')}>${esc(str)}</text>`;
    },

    text: (col, row, str, color, opts) => c.glyphs(c.X(col), c.Y(row), str, color, opts),

    /** Box-drawing and block characters filling the whole cell. */
    tall: (col, row, str, color, opts = {}) =>
      c.glyphs(c.X(col), c.tallY(row), str, color, { ...opts, size: tallSize, squeeze: true }),

    /** Sequence of colored spans: [[text, color, options], ...]. */
    spans(col, row, segments) {
      let cursor = col;
      let out = '';
      for (const [str, color, opts] of segments) {
        out += opts?.tall ? c.tall(cursor, row, str, color, opts) : c.text(cursor, row, str, color, opts);
        cursor += len(str);
      }
      return out;
    },

    /** Cell background. */
    bg: (col, row, widthCols, color, rows = 1) =>
      `<rect x="${c.X(col)}" y="${c.top(row)}" width="${n2(widthCols * cw)}" height="${n2(rows * lh)}" fill="${color}"/>`,

    reveal: (inner, time) => (animate && inner
      ? `<g opacity="0"><set attributeName="opacity" to="1" begin="${c.at(time)}" fill="freeze"/>${inner}</g>`
      : inner),

    /** Visible only between two instants (the end can be Infinity). */
    during(inner, from, to) {
      if (!animate) return to === Infinity ? inner : '';
      const off = to === Infinity ? '' : `<set attributeName="visibility" to="hidden" begin="${c.at(to)}" fill="freeze"/>`;
      return `<g visibility="hidden"><set attributeName="visibility" to="visible" begin="${c.at(from)}" fill="freeze"/>${off}${inner}</g>`;
    },

    /** Horizontal clip that grows cell by cell (typing, bars). */
    grow(inner, col, row, cells, time, stepDur, rows = 1) {
      if (!animate || cells <= 0) return { svg: inner, end: time };
      const id = c.id('g');
      const values = Array.from({ length: cells + 1 }, (_, i) => (i ? n2(i * cw + 3) : 0)).join(';');
      defs.push(
        `<clipPath id="${id}"><rect x="${n2(c.X(col) - 1)}" y="${c.top(row)}" height="${n2(rows * lh)}" width="0">` +
        `<animate attributeName="width" values="${values}" begin="${c.at(time)}" dur="${c.at(cells * stepDur)}" calcMode="discrete" fill="freeze"/></rect></clipPath>`,
      );
      return { svg: `<g clip-path="url(#${id})">${inner}</g>`, end: time + cells * stepDur };
    },

    typed(col, row, str, color, time, perChar, opts) {
      const node = c.text(col, row, str, color, opts);
      return c.grow(node, col, row, len(str), time, perChar);
    },

    /** Block cursor; blinks from `time` on. */
    cursor(col, row, time, color = theme.cursor) {
      const rect = `x="${c.X(col)}" y="${n2(c.top(row) + 2)}" width="${n2(cw)}" height="${n2(lh - 4)}" fill="${color}"`;
      return animate
        ? `<rect ${rect} opacity="0"><animate attributeName="opacity" values="1;0" dur="1.1s" begin="${c.at(time)}" calcMode="discrete" repeatCount="indefinite"/></rect>`
        : `<rect ${rect}/>`;
    },

    /** Progress bar with 1/8-cell precision. */
    blockBar(col, row, cells, fraction, color, time, stepDur, track = theme.border) {
      const eighths = Math.round(Math.max(0, Math.min(1, fraction)) * cells * 8);
      const full = Math.floor(eighths / 8);
      const part = EIGHTHS[eighths % 8];
      const filled = '█'.repeat(full) + part;
      const used = full + (part ? 1 : 0);
      const trackSvg = c.text(col + used, row, '░'.repeat(Math.max(0, cells - used)), track);
      const grown = c.grow(c.text(col, row, filled, color, { seal: true }), col, row, used, time, stepDur);
      return { svg: trackSvg + grown.svg, end: grown.end };
    },

    /**
     * Text strip scrolling inside a window of `windowCols` cells.
     * `strip` needs `steps + windowCols` characters for a seamless loop.
     */
    ticker(col, row, strip, windowCols, color, steps, stepDur, time, opts = {}) {
      const node = (x) => (opts.tall
        ? c.glyphs(x, c.tallY(row), strip, color, { ...opts, size: tallSize, squeeze: true })
        : c.glyphs(x, c.Y(row), strip, color, opts));
      if (!animate) return `<g>${c.clipWindow(col, row, windowCols, 1, node(c.X(col)))}</g>`;
      const values = Array.from({ length: steps }, (_, i) => n2(c.X(col) - i * cw)).join(';');
      const moving = node(c.X(col)).replace('<text ', `<text data-t="1" `)
        .replace('</text>', `<animate attributeName="x" values="${values}" begin="${c.at(time)}" dur="${c.at(steps * stepDur)}" calcMode="discrete" repeatCount="indefinite"/></text>`);
      return c.clipWindow(col, row, windowCols, 1, moving);
    },

    clipWindow(col, row, widthCols, rows, inner) {
      const id = c.id('w');
      defs.push(`<clipPath id="${id}"><rect x="${c.X(col)}" y="${c.top(row)}" width="${n2(widthCols * cw)}" height="${n2(rows * lh)}"/></clipPath>`);
      return `<g clip-path="url(#${id})">${inner}</g>`;
    },
  };
  return c;
}

File: scripts/terminal/svg.py
"""
Drawing primitives on a grid of terminal cells.

Each grid row is `lh` tall. Regular text uses the font at its normal size, with
the baseline centered in the cell. Box-drawing and block characters (╭─│█▄)
use "tall" mode: the font is scaled up until the glyph fills the whole cell and
squeezed horizontally, so lines and blocks connect without gaps between rows,
like in a real terminal.
"""
import math

from .fmt import INF, esc, jround, n2, num

EIGHTHS = ["", "▏", "▎", "▍", "▌", "▋", "▊", "▉"]


class Canvas:
    def __init__(self, theme, layout, animate, k, glow):
        self.T = theme
        self.fs = layout["fontSize"]
        self.cw = self.fs * 0.6
        self.lh = n2(self.fs * layout["lineHeight"])
        self.padX = 26
        self.titleH = 36
        self.originY = self.titleH + 16
        self.cols = math.floor((layout["width"] - self.padX * 2) / self.cw)
        self.tallSize = n2(self.lh / 1.32 + 0.15)
        self.width = layout["width"]
        self.animate = animate
        self.k = k
        self._glow = glow
        self.defs = []
        self._uid = 0

    # ── coordinates ──
    def id(self, prefix="i"):
        value = f"{prefix}{self._uid}"
        self._uid += 1
        return value

    def X(self, col):
        return n2(self.padX + col * self.cw)

    def top(self, row):
        return n2(self.originY + row * self.lh)

    def Y(self, row):
        return n2(self.originY + row * self.lh + (self.lh + 0.72 * self.fs) / 2)

    def tallY(self, row):
        return n2(self.originY + row * self.lh + 1.02 * self.tallSize - 0.08)

    def at(self, s):
        return f"{num(n2(s))}s"

    # ── drawing ──
    def glow(self, inner):
        """Soft phosphor glow around the content (when enabled)."""
        return f'<g filter="url(#glow)">{inner}</g>' if self._glow and inner else inner

    def glyphs(self, x, y, s, color, size=None, weight=None, anchor=None, squeeze=False,
               opacity=None, seal=False, **_):
        s = str(s)
        n = len(s)
        if not n or not s.strip():
            return ""
        size = self.fs if size is None else size
        attrs = [f'x="{num(n2(x))}"', f'y="{num(n2(y))}"', f'fill="{color}"']
        advance = self.cw if squeeze else size * 0.6
        if n > 1 or squeeze:
            attrs.append(f'textLength="{num(n2(n * advance))}"')
            attrs.append(f'lengthAdjust="{"spacingAndGlyphs" if squeeze else "spacing"}"')
        if size != self.fs:
            attrs.append(f'style="font-size:{num(size)}px"')
        if weight:
            attrs.append(f'font-weight="{weight}"')
        if anchor:
            attrs.append(f'text-anchor="{anchor}"')
        if opacity is not None:
            attrs.append(f'fill-opacity="{num(opacity)}"')
        if seal:
            # a thin same-color stroke closes antialiasing gaps between blocks
            attrs.append(f'stroke="{color}"')
            attrs.append('stroke-width="0.6"')
        return f"<text {' '.join(attrs)}>{esc(s)}</text>"

    def text(self, col, row, s, color, **opts):
        return self.glyphs(self.X(col), self.Y(row), s, color, **opts)

    def tall(self, col, row, s, color, **opts):
        """Box-drawing and block characters filling the whole cell."""
        opts = {**opts, "size": self.tallSize, "squeeze": True}
        return self.glyphs(self.X(col), self.tallY(row), s, color, **opts)

    def spans(self, col, row, segments):
        """Sequence of colored spans: [(text, color, {options}), ...]."""
        cursor = col
        out = ""
        for seg in segments:
            s, color = seg[0], seg[1]
            opts = seg[2] if len(seg) > 2 and seg[2] else {}
            out += self.tall(cursor, row, s, color, **opts) if opts.get("tall") else self.text(cursor, row, s, color, **opts)
            cursor += len(str(s))
        return out

    def bg(self, col, row, width_cols, color, rows=1):
        """Cell background."""
        return (f'<rect x="{num(self.X(col))}" y="{num(self.top(row))}" width="{num(n2(width_cols * self.cw))}" '
                f'height="{num(n2(rows * self.lh))}" fill="{color}"/>')

    # ── animation ──
    def reveal(self, inner, time):
        if self.animate and inner:
            return f'<g opacity="0"><set attributeName="opacity" to="1" begin="{self.at(time)}" fill="freeze"/>{inner}</g>'
        return inner

    def during(self, inner, start, end):
        """Visible only between two instants (the end can be INF)."""
        if not self.animate:
            return inner if end == INF else ""
        off = "" if end == INF else f'<set attributeName="visibility" to="hidden" begin="{self.at(end)}" fill="freeze"/>'
        return (f'<g visibility="hidden"><set attributeName="visibility" to="visible" begin="{self.at(start)}" fill="freeze"/>'
                f"{off}{inner}</g>")

    def grow(self, inner, col, row, cells, time, step, rows=1):
        """Horizontal clip that grows cell by cell (typing, bars). Returns (svg, end)."""
        if not self.animate or cells <= 0:
            return inner, time
        cid = self.id("g")
        values = ";".join(num(n2(i * self.cw + 3)) if i else "0" for i in range(cells + 1))
        self.defs.append(
            f'<clipPath id="{cid}"><rect x="{num(n2(self.X(col) - 1))}" y="{num(self.top(row))}" '
            f'height="{num(n2(rows * self.lh))}" width="0">'
            f'<animate attributeName="width" values="{values}" begin="{self.at(time)}" dur="{self.at(cells * step)}" '
            f'calcMode="discrete" fill="freeze"/></rect></clipPath>'
        )
        return f'<g clip-path="url(#{cid})">{inner}</g>', time + cells * step

    def typed(self, col, row, s, color, time, per_char, **opts):
        node = self.text(col, row, s, color, **opts)
        return self.grow(node, col, row, len(str(s)), time, per_char)

    def cursor(self, col, row, time, color=None):
        """Block cursor; blinks from `time` on."""
        color = color or self.T["cursor"]
        rect = (f'x="{num(self.X(col))}" y="{num(n2(self.top(row) + 2))}" width="{num(n2(self.cw))}" '
                f'height="{num(n2(self.lh - 4))}" fill="{color}"')
        if self.animate:
            return (f'<rect {rect} opacity="0"><animate attributeName="opacity" values="1;0" dur="1.1s" '
                    f'begin="{self.at(time)}" calcMode="discrete" repeatCount="indefinite"/></rect>')
        return f"<rect {rect}/>"

    def block_bar(self, col, row, cells, fraction, color, time, step, track=None):
        """Progress bar with 1/8-cell precision. Returns (svg, end)."""
        track = track or self.T["border"]
        eighths = jround(max(0, min(1, fraction)) * cells * 8)
        full = eighths // 8
        part = EIGHTHS[eighths % 8]
        filled = "█" * full + part
        used = full + (1 if part else 0)
        track_svg = self.text(col + used, row, "░" * max(0, cells - used), track)
        grown, end = self.grow(self.text(col, row, filled, color, seal=True), col, row, used, time, step)
        return track_svg + grown, end

    def ticker(self, col, row, strip, window_cols, color, steps, step, time, **opts):
        """Text strip scrolling inside a window of `window_cols` cells."""
        def node(x):
            if opts.get("tall"):
                return self.glyphs(x, self.tallY(row), strip, color, **{**opts, "size": self.tallSize, "squeeze": True})
            return self.glyphs(x, self.Y(row), strip, color, **opts)

        if not self.animate:
            return f"<g>{self.clip_window(col, row, window_cols, 1, node(self.X(col)))}</g>"
        values = ";".join(num(n2(self.X(col) - i * self.cw)) for i in range(steps))
        moving = node(self.X(col)).replace("<text ", '<text data-t="1" ', 1).replace(
            "</text>",
            f'<animate attributeName="x" values="{values}" begin="{self.at(time)}" dur="{self.at(steps * step)}" '
            f'calcMode="discrete" repeatCount="indefinite"/></text>', 1)
        return self.clip_window(col, row, window_cols, 1, moving)

    def clip_window(self, col, row, width_cols, rows, inner):
        cid = self.id("w")
        self.defs.append(
            f'<clipPath id="{cid}"><rect x="{num(self.X(col))}" y="{num(self.top(row))}" '
            f'width="{num(n2(width_cols * self.cw))}" height="{num(n2(rows * self.lh))}"/></clipPath>'
        )
        return f'<g clip-path="url(#{cid})">{inner}</g>'
